File: ring-android/app/src/main/java/cx/ring/tv/account/TVSettingsFragment.java
/*
 * Copyright (C) 2004-2019 Savoir-faire Linux Inc.
 *
 * Author: Pierre Duchemin <pierre.duchemin@savoirfairelinux.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */
package cx.ring.tv.account;

import android.content.Context;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.leanback.preference.LeanbackSettingsFragmentCompat;
import androidx.preference.ListPreference;
import androidx.preference.PreferenceFragmentCompat;
import androidx.preference.PreferenceManager;
import androidx.preference.SwitchPreference;
import androidx.preference.Preference;
import androidx.preference.PreferenceScreen;

import android.view.View;

import java.util.ArrayList;
import java.util.Arrays;

import cx.ring.R;
import cx.ring.application.JamiApplication;
import cx.ring.fragments.GeneralAccountPresenter;
import cx.ring.fragments.GeneralAccountView;
import cx.ring.model.Account;
import cx.ring.model.ConfigKey;
import cx.ring.services.SharedPreferencesServiceImpl;
import cx.ring.utils.Tuple;

public class TVSettingsFragment extends LeanbackSettingsFragmentCompat {

    @Override
    public void onPreferenceStartInitialScreen() {
        startPreferenceFragment(PrefsFragment.newInstance());
    }

    @Override
    public boolean onPreferenceStartFragment(PreferenceFragmentCompat preferenceFragment, Preference preference) {
        return false;
    }

    @Override
    public boolean onPreferenceStartScreen(PreferenceFragmentCompat caller, PreferenceScreen pref) {
        final Fragment prefsFragment = PrefsFragment.newInstance();
        final Bundle args = new Bundle();
        args.putString(PreferenceFragmentCompat.ARG_PREFERENCE_ROOT, pref.getKey());
        prefsFragment.setArguments(args);
        startPreferenceFragment(prefsFragment);
        return true;
    }

    public static class PrefsFragment extends JamiPreferenceFragment<GeneralAccountPresenter> implements GeneralAccountView {
        private boolean autoAnswer;

        public static PrefsFragment newInstance() {
            return new PrefsFragment();
        }

        @Override
        public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
            ((JamiApplication) requireActivity().getApplication()).getInjectionComponent().inject(this);
            super.onViewCreated(view, savedInstanceState);
            presenter.init();
        }

        @Override
        public void addJamiPreferences(String accountId) {

        }

        @Override
        public void addSipPreferences() {

        }

        @Override
        public void accountChanged(@NonNull Account account) {
            // load information from account to ui
            autoAnswer = account.getConfig().getBool(ConfigKey.ACCOUNT_AUTOANSWER);

            SwitchPreference pref = findPreference(ConfigKey.ACCOUNT_AUTOANSWER.key());
            if (pref != null)
                pref.setChecked(autoAnswer);
        }

        @Override
        public void finish() {
            requireActivity().onBackPressed();
        }

        @Override
        public void updateResolutions(Tuple<Integer, Integer> maxResolution) {
            String[] videoResolutionsNames = getResources().getStringArray(R.array.video_resolutionStrings);
            String[] videoResolutionsValues = getResources().getStringArray(R.array.video_resolutions);

            int currentResolution = presenter.getPreferenceService().getResolution();
            videoResolutionsValues = filterResolutions(videoResolutionsValues, currentResolution, maxResolution);

            ListPreference lpVideoResolution = findPreference("video_resolution");
            if (lpVideoResolution != null) {
                lpVideoResolution.setEntries(Arrays.copyOfRange(videoResolutionsNames, 0, videoResolutionsValues.length));
                lpVideoResolution.setEntryValues(videoResolutionsValues);
            }
        }

        @Override
        public void onCreatePreferences(Bundle bundle, String rootKey) {
            PreferenceManager pm = getPreferenceManager();
            pm.setSharedPreferencesMode(Context.MODE_PRIVATE);
            pm.setSharedPreferencesName(SharedPreferencesServiceImpl.PREFS_VIDEO);
            setPreferencesFromResource(R.xml.tv_account_general_pref, rootKey);
        }

        private String[] filterResolutions(String[] videoResolutionsValues, int currentResolution, Tuple<Integer, Integer> maxResolution) {
            if (maxResolution == null) return videoResolutionsValues;
            if (currentResolution > maxResolution.second) return videoResolutionsValues;

            ArrayList<String> resolutions = new ArrayList<>();
            for (String videoResolutionsValue : videoResolutionsValues) {
                int resolutionValueInt = Integer.parseInt(videoResolutionsValue);
                if (resolutionValueInt <= maxResolution.second) {
                    resolutions.add(videoResolutionsValue);
                }
            }

            return resolutions.toArray(new String[0]);
        }

        @Override
        public boolean onPreferenceTreeClick(Preference preference) {
            if (preference.getKey().equals(ConfigKey.ACCOUNT_AUTOANSWER.key())) {
                presenter.twoStatePreferenceChanged(ConfigKey.ACCOUNT_AUTOANSWER, !autoAnswer);
                this.autoAnswer = !autoAnswer;
            }
            return super.onPreferenceTreeClick(preference);
        }
    }
}

File: jami-android/app/src/main/java/cx/ring/tv/settings/TVSettingsFragment.kt
/*
 *  Copyright (C) 2004-2025 Savoir-faire Linux Inc.
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <https://www.gnu.org/licenses/>.
 */
package cx.ring.tv.settings

import android.content.Context
import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.appcompat.app.AlertDialog
import androidx.leanback.preference.LeanbackSettingsFragmentCompat
import androidx.preference.*
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import cx.ring.R
import cx.ring.services.SharedPreferencesServiceImpl
import cx.ring.tv.account.JamiPreferenceFragment
import dagger.hilt.android.AndroidEntryPoint
import net.jami.model.Account
import net.jami.model.ConfigKey
import net.jami.settings.GeneralAccountPresenter
import net.jami.settings.GeneralAccountView
import java.util.*

@AndroidEntryPoint
class TVSettingsFragment : LeanbackSettingsFragmentCompat() {
    override fun onPreferenceStartInitialScreen() {
        startPreferenceFragment(PrefsFragment.newInstance())
    }

    override fun onPreferenceStartFragment(preferenceFragment: PreferenceFragmentCompat, preference: Preference): Boolean {
        val args = preference.extras
        val f = childFragmentManager.fragmentFactory.instantiate(requireContext().classLoader, preference.fragment!!).apply {
            arguments = args
            setTargetFragment(preferenceFragment, 0)
        }
        if (f is PreferenceFragmentCompat
            || f is PreferenceDialogFragmentCompat
        ) {
            startPreferenceFragment(f)
        } else {
            startImmersiveFragment(f)
        }
        return true
    }

    override fun onPreferenceStartScreen(caller: PreferenceFragmentCompat, pref: PreferenceScreen): Boolean {
        startPreferenceFragment(PrefsFragment.newInstance().apply {
            arguments = Bundle().apply {
                putString(PreferenceFragmentCompat.ARG_PREFERENCE_ROOT, pref.key)
            }
        })
        return true
    }

    @AndroidEntryPoint
    class PrefsFragment : JamiPreferenceFragment<GeneralAccountPresenter>(), GeneralAccountView {
        override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
            super.onViewCreated(view, savedInstanceState)
            presenter.init()
        }

        override fun addJamiPreferences(accountId: String) {}
        override fun addSipPreferences() {}
        override fun accountChanged(account: Account) {
            findPreference<SwitchPreference>(ConfigKey.ACCOUNT_AUTOANSWER.key)?.isChecked = account.config.getBool(ConfigKey.ACCOUNT_AUTOANSWER)
            findPreference<SwitchPreference>(ConfigKey.ACCOUNT_ISRENDEZVOUS.key)?.isChecked = account.config.getBool(ConfigKey.ACCOUNT_ISRENDEZVOUS)
        }

        override fun finish() {
            requireActivity().onBackPressedDispatcher.onBackPressed()
        }

        override fun updateResolutions(maxResolution: Pair<Int, Int>?, currentResolution: Int) {
            val videoResolutionsNames = resources.getStringArray(R.array.video_resolutionStrings)
            val videoResolutionsValues =
                filterResolutions(resources.getStringArray(R.array.video_resolutions), currentResolution, maxResolution)
            findPreference<ListPreference>(SharedPreferencesServiceImpl.PREF_RESOLUTION)?.apply {
                entries = videoResolutionsNames.copyOfRange(0, videoResolutionsValues.size)
                entryValues = videoResolutionsValues
            }
        }

        override fun onCreatePreferences(bundle: Bundle?, rootKey: String?) {
            preferenceManager?.apply {
                sharedPreferencesMode = Context.MODE_PRIVATE
                sharedPreferencesName = SharedPreferencesServiceImpl.PREFS_VIDEO
            }
            setPreferencesFromResource(R.xml.tv_account_general_pref, rootKey)
        }

        private fun filterResolutions(
            videoResolutionsValues: Array<String>,
            currentResolution: Int,
            maxResolution: Pair<Int, Int>?
        ): Array<String> {
            if (maxResolution == null) return videoResolutionsValues
            if (currentResolution > maxResolution.second) return videoResolutionsValues
            val resolutions = ArrayList<String>()
            for (videoResolutionsValue in videoResolutionsValues) {
                if (videoResolutionsValue.toInt() <= maxResolution.second)
                    resolutions.add(videoResolutionsValue)
            }
            return resolutions.toTypedArray()
        }

        override fun onPreferenceTreeClick(preference: Preference): Boolean {
            if (preference.key == PREF_DELETE_ACCOUNT) {
                showDeleteAccountDialog()
                return true
            }
            val key = ConfigKey.fromString(preference.key)
            if (key != null && key.isBool)
                presenter.twoStatePreferenceChanged(key, (preference as SwitchPreference).isChecked)
            return super.onPreferenceTreeClick(preference)
        }

        private fun showDeleteAccountDialog() {
            val dialog = MaterialAlertDialogBuilder(requireContext(), com.google.android.material.R.style.Theme_MaterialComponents_Dialog)
                .setTitle(R.string.account_delete_dialog_title)
                .setMessage(R.string.account_delete_dialog_message)
                .setPositiveButton(R.string.menu_delete) { _, _ -> deleteAccount() }
                .setNegativeButton(android.R.string.cancel, null)
                .create()
            dialog.setOnShowListener {
                dialog.getButton(AlertDialog.BUTTON_POSITIVE)?.apply {
                    isFocusable = true
                    isFocusableInTouchMode = true
                    requestFocus()
                }
            }
            dialog.show()
        }

        private fun deleteAccount() {
            presenter.removeAccount()
            val activity = requireActivity()
            Log.d("pavan", "settings calling dispatycher")
            activity.onBackPressedDispatcher.onBackPressed()
            activity.finish()
        }

        companion object {
            private const val PREF_DELETE_ACCOUNT = "Account.delete"
            fun newInstance(): PrefsFragment = PrefsFragment()
        }
    }
}